webapp/components/utils/UniqueSlugForm.spec.js

```diff
@@ -0,0 +1,80 @@
+import UniqueSlugForm from './UniqueSlugForm'
+import Schema from 'async-validator'
+
+let translate
+let apollo
+let currentUser
+
+beforeEach(() => {
+  translate = jest.fn(() => 'Validation error')
+  apollo = {
+    query: jest.fn().mockResolvedValue({ data: { User: [] } }),
+  }
+  currentUser = null
+})
+
+describe('UniqueSlugForm', () => {
+  let validate = object => {
+    const { formSchema } = UniqueSlugForm({ translate, apollo, currentUser })
+    const validator = new Schema(formSchema)
+    return validator.validate(object, { suppressWarning: true }).catch(({ errors }) => {
+      throw new Error(errors[0].message)
+    })
+  }
+
+  describe('regex', () => {
+    describe('non URL-safe characters, e.g. whitespaces', () => {
+      it('rejects', async () => {
+        await expect(validate({ slug: 'uh oh' })).rejects.toThrow('Validation error')
+      })
+    })
+
+    describe('alphanumeric, hyphens or underscores', () => {
+      it('validates', async () => {
+        await expect(validate({ slug: '_all-right_' })).resolves.toBeUndefined()
+      })
+    })
+  })
+
+  describe('given a currentUser with a slug', () => {
+    beforeEach(() => {
+      currentUser = { slug: 'current-user' }
+    })
+
+    describe('backend returns no user for given slug', () => {
+      beforeEach(() => {
+        apollo.query.mockResolvedValue({
+          data: { User: [] },
+        })
+      })
+
+      it('validates', async () => {
+        await expect(validate({ slug: 'slug' })).resolves.toBeUndefined()
+      })
+    })
+
+    describe('backend returns user', () => {
+      let slug
+      beforeEach(() => {
+        slug = 'already-taken'
+        apollo.query.mockResolvedValue({
```

Comment:
> [<img alt="Tirokk" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/Tirokk) **Authored by [Tirokk](https://github.com/Tirokk)**
_<time datetime="2019-09-23T11:12:43Z" title="Monday, September 23rd 2019, 1:12:43 pm +02:00">Sep 23, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -0,0 +1,80 @@
+import UniqueSlugForm from './UniqueSlugForm'
+import Schema from 'async-validator'
+
+let translate
+let apollo
+let currentUser
+
+beforeEach(() => {
+  translate = jest.fn(() => 'Validation error')
+  apollo = {
+    query: jest.fn().mockResolvedValue({ data: { User: [] } }),
+  }
+  currentUser = null
+})
+
+describe('UniqueSlugForm', () => {
+  let validate = object => {
+    const { formSchema } = UniqueSlugForm({ translate, apollo, currentUser })
+    const validator = new Schema(formSchema)
+    return validator.validate(object, { suppressWarning: true }).catch(({ errors }) => {
+      throw new Error(errors[0].message)
+    })
+  }
+
+  describe('regex', () => {
+    describe('non URL-safe characters, e.g. whitespaces', () => {
+      it('rejects', async () => {
+        await expect(validate({ slug: 'uh oh' })).rejects.toThrow('Validation error')
+      })
+    })
+
+    describe('alphanumeric, hyphens or underscores', () => {
+      it('validates', async () => {
+        await expect(validate({ slug: '_all-right_' })).resolves.toBeUndefined()
+      })
+    })
+  })
+
+  describe('given a currentUser with a slug', () => {
+    beforeEach(() => {
+      currentUser = { slug: 'current-user' }
+    })
+
+    describe('backend returns no user for given slug', () => {
+      beforeEach(() => {
+        apollo.query.mockResolvedValue({
+          data: { User: [] },
+        })
+      })
+
+      it('validates', async () => {
+        await expect(validate({ slug: 'slug' })).resolves.toBeUndefined()
+      })
+    })
+
+    describe('backend returns user', () => {
+      let slug
+      beforeEach(() => {
+        slug = 'already-taken'
+        apollo.query.mockResolvedValue({
```


This is how you can avoid intermittent test failings !!!
Great ! 🤗😎